### PR TITLE
SMOK-42791 | Reorder and rename customUIActions elements

### PR DESCRIPTION
**Pull request by krbg**
This allows the user to specify an order to his registered commands. By doing this, we won't have the weird arbitrary order in the Shotgun submenu.

This will also allow to specify special (but optional to ensure backward compatibility) display_name to be able to specify what is the name that Flame should display.

From unordered to ordered

Only by changing https://github.com/shotgunsoftware/tk-config-flameplugin/blob/master/env/project.yml#L73

### CODE REVIEW

File: info.yml [hunk 1]
```diff
@@ -70,18 +70,20 @@ configuration:
...
+                     If name is '' then all commands from the given app instance are started.
+                     If display_name is not present, name will be used instead"
+        allows_empty: True
         default_value: []
         values:
             type: dict
             items:
                 name: { type: str }
                 app_instance: { type: str }
+               #display_name: { type: str } (optional)
```
Comment (jpbrault) [inline info.yml]: Should this line be commented?

File: engine.py [hunk 2]
```diff
@@ -207,22 +207,27 @@ def set_install_root(self, install_root):
...
+                 Each tuple will be in the form (instance_name, display_name, command_name, callback)
         """
         # return a dictionary grouping all the commands by instance name
         commands_by_instance = {}
         for (name, value) in self.commands.iteritems():
             app_instance = value["properties"].get("app")
-            if app_instance is None:
-                continue
-            instance_name = app_instance.instance_name
+            if app_instance:
```
Comment (jfboismenu) [inline engine.py]: I don't think null is a valid instance name. Why did you need to do this?

File: flame_hooks/sg_hook.py [hunk 3]
```diff
@@ -63,29 +63,31 @@ def getMainMenuCustomUIActions( ):
...
     # omit system actions that are on the context menu
-    for command_name in engine.commands:
-        properties = engine.commands[command_name]["properties"]
-        if command_name not in commands and properties.get("type") != "context_menu":
-            commands.append(command_name)
+    context_command_names = [context_command[0] for context_command in context_commands]
+    for engine_command_name in engine.commands:
+        properties = engine.commands[engine_command_name]["properties"]
+        if engine_command_name not in context_command_names and properties.get("type") != "context_menu":
+            context_commands.append((engine_command_name, engine_command_name))
```
Comment (jfboismenu) [inline flame_hooks/sg_hook.py]: Looking at the previous logic and yours, it seems you are forgetting to update the context_command_names array with new commands as you insert them.

File: engine.py (hunk 2 above)
Comment (krbg) [inline engine.py]: Because commands that get registered by the tk-plugin-flame bootstrap don't have any instance name so when I want to specify these commands [here](https://github.com/shotgunsoftware/tk-config-flameplugin/blob/b7e09964785a25d6f7bf89ebe850a7747f6f917f/env/project.yml#L73-L75), I need a dummy instance_name.

File: info.yml (hunk 1 above)
Comment (krbg) [inline info.yml]: Yes otherwise this value will be mandatory and it will break the backward compatibility

Comment (jfboismenu) [inline info.yml]: Then we should take it out.

File: engine.py (hunk 2 above)
Comment (jfboismenu) [inline engine.py]: Okay, I've just read the code from sgtk/platform/engine.py and I understand what is going on now. I would prefer if you could set this to tk-flame instead so the value makes a bit more sense with the following comment.
```python
# A command without an app instance in the context menu is actually coming from the engine, so we'll
# use the engine name instead.
```
